opp_path_selection/src/path_selection_artist.cpp

```diff
@@ -0,0 +1,515 @@
+  if (!success)
+  {
+    ROS_ERROR("collectPathPointsCloudCb failed");
+    return false;
+  }
+  if (!srv.response.success)  // no points found
+  {
+    points_idx.clear();  // this is not a failure, we allow zero points for the heat method
+    return true;
+  }
+
+  // the response includes res.cloud_indices which represent the path of vertices between each segment of markers/points
+  for (int i = 0; i < srv.response.cloud_indices.size(); i++)
+  {
+    int idx = srv.response.cloud_indices[i];
+    points_idx.push_back(idx);
+  }
+  return true;
+}
+
+bool PathSelectionArtist::collectPathMesh(const shape_msgs::Mesh& mesh_msg,
+                                          std::vector<int>& points_idx,
+                                          std::string& message)
+{
+  // first get the mesh
+  opp_msgs::GetPathSelectionMesh srv;
+  srv.request.input_mesh = mesh_msg;
+
+  // now get the path segments from markers and find point indices from mesh along each segment
+  bool success = collectPathPointsMeshCb(srv.request, srv.response);
+  message = srv.response.message;
+  points_idx.clear();
+  if (!success)
+  {
+    return false;
+  }
+  if (!srv.response.success)
+  {
+    message = "No points found on mesh";
+  }
+
+  // the response includes res.cloud_indices which represent the path of vertices between each segment of markers/points
+  for (int i = 0; i < srv.response.mesh_indices.size(); i++)
+  {
+    int idx = srv.response.mesh_indices[i];
+    points_idx.push_back(idx);
+  }
+
+  return true;
+}
+
+bool PathSelectionArtist::collectPathPointsCloudCb(opp_msgs::GetPathSelectionCloudRequest& req,
+                                                   opp_msgs::GetPathSelectionCloudResponse& res)
+{
+  // get points from existing marker_array_
+  auto points_it = std::find_if(marker_array_.markers.begin(),
+                                marker_array_.markers.end(),
+                                [](const visualization_msgs::Marker& marker) { return marker.id == 0; });
+
+  // Convert the selection points to Eigen vectors
+  std::vector<Eigen::Vector3d> points;
+  for (const geometry_msgs::Point& pt : points_it->points)
+  {
+    Eigen::Vector3d e;
+    tf::pointMsgToEigen(pt, e);
+    points.push_back(std::move(e));
+  }
+
+  if (points.size() < 2)
+  {
+    res.cloud_indices.clear();
+    res.success = false;
+    res.message = "No Points Selected, use Geometric Axis";
+    return true;
+  }
+
+  // get point cloud from service request
+  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>());
+  pcl::fromROSMsg(req.input_cloud, *cloud);
+
+  // get parameters from file. Note: This file is specified on construction, re-read allows dynamic changes
+  PathSelectorParameters params;
+  bool success = message_serialization::deserialize(path_selection_config_file, params);
+  if (!success)
+  {
+    ROS_ERROR_STREAM("Could not load path selection config from: " << path_selection_config_file);
+    return false;
+  }
```

Comment:
If you want dynamic behavior, this should be part of the `GetPathSelectionCloud` service definition. Loading a file (with a static name and location) in the middle of a function is extremely opaque to any user who hasn't read the source code of this function

Reply:
The only parameter read is no longer used. I'll remove this function altogether.